### ISSUE TITLE
Missing `account_exists` implementation prevents execution of sharding transactions

### ISSUE THREAD

**Issue by jannikluhn** [ethereum/py-evm#281]
### What is wrong?

Executing sharding transactions currently does not work with the FlatTrieBackend as it doesn't implement `account_exists`.

Note that it is not related to account creation transactions per se (as I assumed earlier), normal transactions are affected in the same way.

### How can it be fixed

Either implement `account_exists` or don't use it. It is called in some SELFDESTUCT (e.g. https://github.com/ethereum/py-evm/blob/362266c323c83a851f7dff0a2a59796abaa3f318/evm/logic/system.py#L64) and some CALL (e.g. https://github.com/ethereum/py-evm/blob/362266c323c83a851f7dff0a2a59796abaa3f318/evm/logic/call.py#L287) opcodes. It changes from fork to fork though and I'm not sure if this is still needed, especially as at least one one motivation seems to have been the dust clean up in Spurious Dragon. Also, it seems to be used for gas calculation only which will be different with stateless clients anyway.

It is also called indirectly via `touch_account` for every message, but this can unproblematically be replaced by setting nonce/balance/code to zero unless they are non-zero.

Implementing `account_exists` is probably the easier solution (with `check_if_branch_exist` in py-trie), but I don't like this as much. But that's mainly due to me not seeing why such a function would be necessary.

**Comment by hwwhww** [comment]
Checked with V:

* We can remove `SELFDESTRUCT` opcode from sharding now.
* For `CALL` opcode, we may have to override it by adding `CallSharding` class:
    * Only check `account_is_empty`, not `account_exists`.
    * Look over the logic of `CALL`, the `nonce` part may be very different from the account abstraction model?

**Comment by jannikluhn** [comment]
> Look over the logic of CALL, would the nonce part become very different with the account abstraction model?

I was under the impression that contract nonces would be removed altogether. In its current implementation CREATE2 uses the contract nonce though, so I'm not sure about this. But I don't think it matters a lot, the only thing that would be affected is the definition of an "empty account" (`code/balance/nonce = 0` to `code/balance = 0`).

**Comment by NIC619** [comment]
I can get my hands on `CallSharding`, removing `SELFDESTRUCT`, `account_is_empty`, `account_exist` and `touch_account` after `PAYGAS` opcode is done.

**Comment by jannikluhn** [comment]
I don't think the account methods need to be removed, rather changed slightly:

- `account_is_empty` should not check the nonce
- `account_exists` should be kept as is (raising a `NotImplementedError`)
- `touch_account` should not rely on `account_exists`, but have the same effect (just set nonce/balance/code to zero unless they are set already)

**Comment by jannikluhn** [comment]
@pipermerriam I'm not sure what the right way to proceed is here: For main Ethereum we need to store nonces, for sharding we don't. What's the best way to implement this in the two TrieBackends?

I see three options:

  1) Only `NestedTrieBackend` stores nonces, `FlatTrieBackend` doesn't.
  2) Add a flag `store_nonces` to both backends. Check this backend when doing anything nonce related (throw exception when nonces are got/set, set/don't set nonce when touching account, ...)
  3) Add `FlatTrieBackendWithoutNonces` that inherits from FlatTrieBackend and overrides all nonce related methods to raise `NotImplementedError`s.

1\) is definitely easiest and I doubt it will create a lot of problems in practice. But it couples two things that aren't related (*how* state is stored and *what* state is stored). 2) could work, but all these flags may get out of hand at some point (`state_backend_class`, `store_nonces`, maybe `trie_class` in the future, ...). 3) also doesn't appear very elegant.

**Comment by pipermerriam** [comment]
@jannikluhn What about collapsing the abstraction from a single `AccountStateDB` and multiple backends to just have multiple implementations of the `AccountStateDB`, one for `FlatAccountStateDB` and one for `NestedAccountStateDB`.  We can explore if there is a sufficient *common* API to declare a `BaseStateDB` class, and raise `NotImplementedError` in any methods on either class which don't apply, or we can just not have those methods with the expressed knowledge that they aren't available for any VM which uses that version of the account state database?

I dislike the *flag* based approach.  Those types of logic switches are error prone and lead to confusing code internals that are hard to follow in-my-opinion.

**Comment by jannikluhn** [comment]
> What about collapsing the abstraction from a single AccountStateDB and multiple backends to just have multiple implementations of the AccountStateDB, one for FlatAccountStateDB and one for NestedAccountStateDB.

I think you're right, the abstraction isn't too useful right now. Only noteworthy shared functionality is parameter validation.

Instead of calling them `Flat` and `Nested` what about calling them `ShardingAccountStateDB` and `MainAccountStateDB` (or just `AccountStateDB`)? That way we could remove nonce handling and switch to the binary trie with a clear conscience in `ShardingAccountStateDB`.
